KIP-154 KAFKA-4667 Connect uses AdminClient to create internal topics when needed

rhauch opened the pull request:
The backing store for offsets, status, and configs now attempts to use the new AdminClient to look up the internal topics and create them if they don’t yet exist. If the necessary APIs are not available in the connected broker, the stores fall back to the old behavior of relying upon auto-created topics. Kafka Connect requires a minimum of Apache Kafka 0.10.0.1-cp1, and the AdminClient can work with all versions since 0.10.0.0.

All three of Connect’s internal topics are created as compacted topics, and new distributed worker configuration properties control the replication factor for all three topics and the number of partitions for the offsets and status topics; the config topic requires a single partition and does not allow it to be set via configuration. All of these new configuration properties have sensible defaults, meaning users can upgrade without having to change any of the existing configurations. In most situations, existing Connect deployments will have already created the storage topics prior to upgrading.

The replication factor defaults to 3, so anyone running Kafka clusters with fewer nodes than 3 will receive an error unless they explicitly set the replication factor for the three internal topics. This is actually desired behavior, since it signals the users that they should be aware they are not using sufficient replication for production use.

The integration tests use a cluster with a single broker, so they were changed to explicitly specify a replication factor of 1 and a single partition.

The `KafkaAdminClientTest` was refactored to extract a utility for setting up a `KafkaAdminClient` with a `MockClient` for unit tests.

rhauch committed:
KAFKA-4667 Connect can create internal topics using AdminClient

The backing store for offsets, status, and configs now attempts to use the new AdminClient to look up the
internal topics and create them if they don’t yet exist. If the necessary APIs are not available in the connected
broker, the stores fall back to the old behavior of relying upon auto-created topics. Kafka Connect requires
a minimum of Apache Kafka 0.10.0.1-cp1, and the AdminClient can work with all versions since 0.10.0.0.

All three of Connect’s internal topics are created as compacted topics, and new distributed worker configuration
properties control the replication factor for all three topics and the number of partitions for the offsets
and status topics; the config topic requires a single partition and does not allow it to be set via configuration.
All of these new configuration properties have sensible defaults, meaning users can upgrade without having to
change any of the existing configurations. In most situations, existing Connect deployments will have already
created the storage topics prior to upgrading.

The replication factor defaults to 3, so anyone running Kafka clusters with fewer nodes than 3 will receive an
error unless they explicitly set the replication factor for the three internal topics. This is actually desired behavior,
since it signals the users that they should be aware they are not using sufficient replication for production use.

The integration tests use a cluster with a single broker, so they were changed to explictly specify a replication factor of 1
and a single partition.

ijuma commented:
cc @cmccabe for the AdminClient changes.

rhauch commented:
Per @ewencp this will require a KIP, which I'm working on.

rhauch commented:
See [KIP-154](https://cwiki.apache.org/confluence/display/KAFKA/KIP-154+Add+Kafka+Connect+configuration+properties+for+creating+internal+topics).

ewencp commented on config/connect-distributed.properties:
```diff
@@ -37,15 +37,34 @@ internal.value.converter=org.apache.kafka.connect.json.JsonConverter
+# Topic to use for storing offsets. This topic should have many partitions and be replicated and compacted.
+# Kafka Connect will attempt to create the topic automatically when needed, but you can always manually create
+# the topic before starting Kafka Connect if a specific topic configuration is needed.
+# Most users will want to use the built-in default replication factor of 3 or in some cases even specify a larger value.
```
The comments here are great -- should direct people to set them properly. Might want to also update the comment way at the top of the file. It's currently a bit brief. Maybe just one more sentence about a production configuration should be sure to update key configurations such as bootstrap.servers and replication factors.

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -209,14 +239,39 @@
+                        ConfigDef.Importance.LOW,
+                        OFFSET_STORAGE_PARTITIONS_CONFIG_DOC)
+                .define(OFFSET_STORAGE_REPLICATION_FACTOR_CONFIG,
+                        ConfigDef.Type.INT,
```
We seem to have ended up with different types for replication factor in different places -- it looks like it's ultimately forced to be a `short` but we allow `int` here and throughout other code. Is there any reason not to use `short` throughout?

rhauch replied:
I did see that, and wasn't sure what to do. We can use `short`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedConfig.java:
```diff
@@ -209,14 +239,39 @@
                         ConfigDef.Type.STRING,
                         ConfigDef.Importance.HIGH,
                         OFFSET_STORAGE_TOPIC_CONFIG_DOC)
+                .define(OFFSET_STORAGE_PARTITIONS_CONFIG,
```
Should we add validators on these? num partitions and replication factor should always be >= 1.

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -119,22 +143,27 @@ public void start() {
+                partitions.add(new TopicPartition(topic, entry.getValue().partition()));
+            }
+        } else {
+            log.debug("Unable to find or create topic '{}', so falling back to auto-creation.", topic);
```
This is probably good at `INFO` level as it should be a one-time message (unless you can't create the topics, in which case seeing it repeatedly is helpful) and it's critical to normal operation of Connect.

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -119,22 +143,27 @@ public void start() {
         List<TopicPartition> partitions = new ArrayList<>();
+        TopicDescription topicDesc = topicSupplier.getOrCreateTopic(topic);
+        if (topicDesc != null) {
+            for (Map.Entry<Integer, TopicPartitionInfo> entry : topicDesc.partitions().entrySet()) {
```
While we're in here making changes anyway, might be nice to add debug logging indicating what partitions are being added. This will be helpful for debugging issues where the topics are created incorrectly. Maybe just log the `TopicDescription` as its `toString` seems to give all the relevant info.

rhauch replied:
I didn't put any debug messages in `KafkaBasedLog` because at this point it's not clear whether the topics were in fact created or whether they had already existed. Instead, the `TopicAdmin` utility that does all the work has log statements to report whether for every topic whether it was created or found to already exist.

~However, I do think that it's worth adding to this code a similar log message to what's in `TopicAdmin` when it falls back to auto-creation. That happens starting on line 148 and is used only when the `TopicAdmin` can't use the `AdminClient` with the given broker -- should ideally never happen but is a potential condition.~

rhauch replied:
Actually, the `TopicAdmin` method will always run, and it logs what it does even when it can't use the `AdminClient`. That means no additional log messages here should be necessary.

ewencp replied:
Besides knowing if the topic exists or not, it's also handy to have information about the number of partitions, replication factor, etc. Obviously it would be ideal if we could always rely on this code to create the topics with good configurations, but I meant that having that info to identify that, e.g., the user created a config topic with multiple partitions and that's why they'd be seeing weird behavior.

Anyway, it's not critical, just thought it might help with debuggability.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,323 @@
+         * @return this builder to allow methods to be chained; never null
+         */
+        public NewTopicBuilder replicationFactor(int replicationFactor) {
+            this.replicationFactor = (short) replicationFactor;
```
If we do allow an `int` here, we should a) check for bounds and b) re-validate values after the cast. Otherwise you can end up with invalid results (which should never happen in practice given the range of sane values for replication factor, but down casts like this always warrant such checks).

However, I suspect just shifting to using `short` all the way back to the configs is probably a better solution.

rhauch replied:
ack changing to `short`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,323 @@
+         * @return this builder to allow methods to be chained; never null
+         */
+        public NewTopicBuilder compacted() {
+            this.configs.put("cleanup.policy", "compact");
```
Hmm, would be good to get these constants out of here, but seems they don't appear in the clients jar yet. Not something we need to do now, but we might consider this for the future, and the fact that we have the AdminClient in clients might be good enough reason (though since these are really broker configs, I'm not sure how much we can do with them in the AdminClient).

rhauch replied:
+1. I'll at least pull these into constants now.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,323 @@
+    @Override
+    public String toString() {
+        Object servers = adminConfig.get(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG);
+        return servers != null ? servers.toString() : "<unknown>";
```
Include class name here? This seems like it would be confusing if logged.

rhauch replied:
This method is currently used only in log messages, so rather than have a `toString()` that is potentially difficult to understand, it'd be better to get rid of it altogether and deal with the log messages in a different way.

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java:
```diff
@@ -0,0 +1,302 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+
+/*
```
Extra copy of the license

rhauch replied:
ack

ewencp commented on tests/kafkatest/tests/connect/connect_distributed_test.py:
```diff
@@ -45,8 +45,13 @@ class ConnectDistributedTest(Test):
 
     TOPIC = "test"
     OFFSETS_TOPIC = "connect-offsets"
+    OFFSETS_REPLICATION_FACTOR = "1"
```
Should we also make sure we turn off auto-topic creation in at least one of these? The order of operations should ensure we to create the topic first, but always nice to have an end-to-end validation of that case.

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,323 @@
+
+        try {
+            // Check for existing topics
+            Map<String, TopicDescription> descriptions = describeTopics(requestedTopicNames);
```
I think you need to be careful about auto topic creation: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java#L136-L140

Sending a metadata fetch with a topic that doesn't exist will create the topic if auto create is enabled. As suggested in the javadoc, you might want to work around this with a `listTopics()` instead. Definitely not ideal since it gets the full list of topics, but gets us the information we need and at least this only happens a couple of times at startup (although this is one drawback to the current implementation that has the `KafkaBasedLog` generate the topics independently).

rhauch replied:
ack

ewencp commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/KafkaBasedLogTest.java:
```diff
@@ -131,7 +133,7 @@ public void onCompletion(Throwable error, ConsumerRecord<String, String> record)
     public void setUp() throws Exception {
         store = PowerMock.createPartialMock(KafkaBasedLog.class, new String[]{"createConsumer", "createProducer"},
-                TOPIC, PRODUCER_PROPS, CONSUMER_PROPS, consumedCallback, time);
+                TOPIC, PRODUCER_PROPS, CONSUMER_PROPS, consumedCallback, time, TOPIC_SUPPLIER);
```
Probably one test, perhaps just with a mock, that validates the `TOPIC_SUPPLIER` is invoked?

ewencp reviewed (commented):
In addition to the other comments, a minor nit: can you put the KIP # in the PR title?

rhauch committed:
Improved documentation in the distributed configuration

rhauch committed:
Changed tests to disallow auto-created topics

rhauch committed:
Validate partition and replication factor configuration values

Also changed to `short` for replication factor configurations, and moved some magic string values into constants.

rhauch committed:
Removed extra header

rhauch committed:
Changed logic to further prevent possibly auto-creating topics

Using the admin client to describe a set of non-existant topics may result in those topics being auto-created. Instead, Connect’s `TopicAdmin` class now uses the admin client’s list topics method to determine which topics exist, and then creates those that don’t yet exist. Only after missing topics are created does `TopicAdmin` attempt to describe the topics.

This new logic to list existing topics is potentially expensive when large numbers of topics already exist. However, it’s currently the only way to determine whether a set of topics exists without risking auto-creation. And, this is done only a few times during Kafka Connect startup (once for each of the config, status, and offset logs), so this cost is not paid during normal connector operation.

rhauch committed:
Improved the unit test for KafkaBasedLog

Verified that the TopicSupplier is indeed called during normal startup.

rhauch commented:
@ewencp Updated PR with corrections per your previous comments.

rhauch commented on clients/src/test/java/org/apache/kafka/clients/admin/MockKafkaAdminClientEnv.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.HashMap;
+import java.util.Map;
+
+/**
```
@cmccabe here's the `MockKafkaAdminClientEnv` class that for the most part was pulled out from your `KafkaAdminClientTest` class, making it possible to create an `AdminClient` that uses mock clients. See [this unit test](https://github.com/apache/kafka/pull/2984/files#diff-1480282b21623cc8274c96dd72f1acf1R57) for an example of how it's used to test the `TopicAdmin` component that itself uses the `AdminClient`.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -67,6 +69,19 @@
+         */
+        TopicDescription getOrCreateTopic(String topicName);
+    }
+
```
@cmccabe Here's the interface in Kafka Connect that uses the `TopicDescription` class. Code that uses `KafkaBasedLog` will provide their own implementation of `TopicSupplier`, and that way `KafkaBasedLog` can use this when necessary to create topics without having any knowledge of _how_ that is done. The challenge then becomes how to test `KafkaBasedLog`, since the unit test will need to provide `TopicSupplier` implementations that return `TopicDescription` instances; see [here](https://github.com/apache/kafka/pull/2984/files#diff-28145281473ec8c8a817252d6dbb82b9R110). This is where the public constructor on `TopicDescription` (and `TopicPartitionInfo`) come in very handy and simplify things a great deal.

rhauch committed:
Corrected test cases and fixed new TopicAdmin logic

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,368 @@
+            Map<String, TopicListing> topicListingsByName = listingResults.namesToDescriptions().get();
+            for (String topicName : topicNames) {
+                TopicListing listing = topicListingsByName.get(topicName);
+                if (listing != null) {
```
This shouldn't be possible, right? It wouldn't make much sense to put a topic in the result if it didn't have a corresponding `TopicListing`.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,368 @@
+        try {
+            // Getting topic descriptions for non-existant topics may auto-create them, so first see which of
+            // our topics do already exist
+            Set<String> existingTopicNames = existingTopics(topicsByName.keySet());
```
`existingTopics(allRequestedTopicNames)`

rhauch replied:
ack

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,368 @@
+            // Getting topic descriptions for non-existant topics may auto-create them, so first see which of
+            // our topics do already exist
+            Set<String> existingTopicNames = existingTopics(topicsByName.keySet());
+            if (existingTopicNames == null) {
```
All the `null` checks at each layer of the call stack make me think that particular issue might be better handled with an exception.

Not critical since this is all internal code, but seems like then we'd only need to check version compatibility in one or two places.

rhauch replied:
Yeah, I debated using an exception. 

However, I'm rethinking this logic, and think that I can make it more straightforward and potentially push down some of the topic creation logic into the `AdminClient`, getting rid of the Kafka Connect `TopicAdmin` utility, and further simplifying the Kafka Connect changes. 

The new `TopicAdmin` logic tries to be smart about only creating topics that don't yet exist _and_ to then return a description of all of the topics. I don't think it needs to do all this extra work of listing all of the topics, creating the missing topics, and fetching the descriptions of all the topics. Instead, it should just try to create the topics, handle the `TopicExistsException`, and just return the names of the topics that were created. This is simple enough that we could push this logic down into the `AdminClient` on top of what's already there, and get rid of `TopicAdmin` altogether.

Then `KafkaBasedLog` could instead start up by using the `TopicAdmin` to ensure the topics are created, logging whether the topics were created or already exist. The remainder of `KafkaBasedLog` can be what it used to be, including setting up the consumers as before, since the topics should already be created at that point.

I'll put together a commit that makes these changes, and can work with @cmccabe to see if the changes to `AdminClient` would be acceptable.

ewencp commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -0,0 +1,368 @@
+            }
+
+            // Remove the topics that already exist
+            for (String existingTopicName : existingTopicNames) {
```
You can also do this more concisely as `topicsByName.keySet().removeAll(existingTopicNames)`.

rhauch replied:
I agree, except that I wanted a debug statement that printed out the topics that existed. I can rearrange this, though.

ewencp reviewed (commented):
@rhauch A couple of final nits, and we need to sort out the AdminClient stuff.

I'm a bit surprised how much of the patch is just for a more convenient layer on top of the AdminClient... kinda suggests there's something still missing for that layer.

rhauch committed:
Fairly significant refactoring to simplify the use of AdminClient

Simplified the logic of the `TopicAdmin` utility to attempt to create the topics and handle any topics that already exist. Then, the `Kafka*Store` implementations use the `TopicAdmin` utility in a simpler way, providing a custom initialization function to its `KafkaBasedLog` instance. The latter calls the initialization function (to create the topics if necessary) prior to starting the consumer, ensuring that the topics are created before the consumer attempts to use them.

rhauch commented:
As mentioned earlier, I've refactored the changes to make them simpler and less invasive on existing logic in the `KafkaBasedLog`. The `TopicAdmin` internal utility class in Kafka Connect provides a slightly more usable layer on top of `AdminClient`, but its `createTopics` method is much simpler than before and attempts to create a set of topics and return the names of the topics that were in fact created; any topic name not included in the result implies an existing topic. (This same method in earlier commits attempted to list the topics and create only those topics that didn't yet exist, but doing this required extra requests to the brokers for little or no benefit.) 

@cmccabe and @ewencp: it would be possible to move that `TopicAdmin` logic into a new method on `CreateTopicResults` to be something like:

```
    /**
     * Return a future that succeeds after all topic creation requests have been processed.
     * @return a future that returns the names of all topics that were newly created by the request
     */
    public KafkaFuture<Set<String>> createdTopicNames() {
        ...
    }
```

So while I do think that this would be valuable, it doesn't seem to have the same API feel as the rest of the `AdminClient` API. And since Kafka Connect needs to perform this logic in multiple places, I've chosen to put that logic into the `TopicAdmin` internal utility class in Kafka Connect.

Let me know if you think it would be better to have on the admin client's `CreateTopicResults`, or if there are any other suggestions/corrections.

rhauch commented:
@ewencp @cmccabe any thoughts on this updated PR?

ewencp commented:
@rhauch This looks ready to go but 9815e18fefa27d8d901356ec1d994a41b4db4622 modified the admin client tests further. Can you clean up and then I can commit?

rhauch commented:
@ewencp rebased on `trunk` and updated this PR.

ewencp commented:
LGTM
